### PR TITLE
f(x): GoogleMapsGeocoder.new accepts optional Logger

### PR DESCRIPTION
**Pull request by ivanoblomov**
# Closes: #138

## Goal
Expose the gem's logger so users can customize it.

## Approach
1. Change `GoogleMapsGeocoder.new` to accept an optional `Logger`.
2. Default 1 to standard error.
3. Subclass `GeocodingError` from the more specific `RuntimeError`.
4. Pass logger to 3.

Signed-off-by: Roderick Monje <rod@foveacentral.com>

### CODE REVIEW

File: lib/google_maps_geocoder/google_maps_geocoder.rb [hunk 1]
```diff
@@ -66,20 +66,19 @@ class GoogleMapsGeocoder
   # object.
   #
   # @param address [String] a geocodable address
+  # @param logger [Logger] a standard Logger, defaults to standard error
   # @return [GoogleMapsGeocoder] the Google Maps result for the specified
   #   address
   # @example
   #   chez_barack = GoogleMapsGeocoder.new '1600 Pennsylvania DC'
-  def initialize(address)
+  def initialize(address, logger = Logger.new($stderr))
```
Comment (olleolleolle) [inline lib/google_maps_geocoder/google_maps_geocoder.rb:74]: I'd make this argument with a default a keyword argument.

**Comment by ivanoblomov** [comment]
Ready for review.

**Review by olleolleolle — DISMISSED** [review]
I had to read the message with the ".new" mention two times, but it's at least descriptive of where it has happened.

**Comment by ivanoblomov** [comment]
Yeah, I've always included the method name when logging an error as a matter of convention. Figure that combined with the `raise` should be enough context. Open to suggestions though.

**Comment by ivanoblomov** [comment]
My bad, one more change. Forgot to pass the logger to the error: efdf8a17e1ea0807a344aafcb0455aaa3ee22e3e

File: lib/google_maps_geocoder/google_maps_geocoder.rb [hunk 2]
```diff
@@ -75,7 +75,7 @@ def initialize(address, logger: Logger.new($stderr))
     @json = address.is_a?(String) ? google_maps_response(address) : address
     status = @json && @json['status']
     raise RuntimeError if status == 'OVER_QUERY_LIMIT'
-    raise GeocodingError, @json if !@json || @json.empty? || status != 'OK'
+    raise GeocodingError.new(@json, logger: logger) if !@json || @json.empty? || status != 'OK'
```
Comment (olleolleolle) [inline lib/google_maps_geocoder/google_maps_geocoder.rb:78]: https://blog.saeloun.com/2021/09/28/ruby-allow-value-omission-in-hash-literals/

```suggestion
    raise GeocodingError.new(@json, logger:) if !@json || @json.empty? || status != 'OK'
```

Comment (ivanoblomov) [reply lib/google_maps_geocoder/google_maps_geocoder.rb]: Right on. I still haven't grokked the new arg syntax.

**Review by olleolleolle — APPROVED** [review]

**Pull request merged**